### ISSUE TITLE
Migrating from slang-torch to slangpy - missing `DiffTensorView` equivalence

### ISSUE THREAD

**Issue by l-oneil-arm** [shader-slang/slangpy#610]
Hi, 

This issue submission was prompted from a discussion on discord around the transition from slang-torch to slangpy. At current, we are looking for a way to move our code base across to slangpy, however, it does not yet seem easily possible, as slangpy is missing something equivalent to the `DiffTensorView` in slang-torch. 

Here's a copy paste of my question in discord that prompted this issue:

> My starting question, is when moving from slangtorch to slangpy, with a shader like this:
>```
> [AutoPyBindCUDA]
> [CUDAKernel]
> [Differentiable]
> void square(DiffTensorView input, DiffTensorView output)
> {
>     uint3 dispatchIdx = cudaThreadIdx() + cudaBlockIdx() * cudaBlockDim();
> 
>     if (dispatchIdx.x >= input.size(0))
>         return;
> 
>     output[dispatchIdx.x] = input[dispatchIdx.x] * input[dispatchIdx.x];
> }
>```
> 
> 
> How do I do this with the fewest possible shader changes? e.g., what is the equivalent type I should use instead of DiffTensorView, which doesn't seem to exist in slangpy?
> 
> I'm not looking for a kernel that looks like this:
>```
> [Differentiable]
> float square(float x) {
>     return x * x;
> }
>```
> 
> Because in some shaders I perform e.g., linear interpolation etc., which requires loading from multiple locations per-thread, and gradient propagation to multiple taps. 
> 
> What is your suggestion?
> 

I look forward to discussion, thanks in advance.
Liam

**Comment by ccummingsNV** [comment]
Hi. The closest we have at the moment is the GradInTensor (for when you want to pass in gradients), the GradOutTensor (for passing out gradients) and the GradInOutTensor (for situations where you're both reading 1 set of gradients and writing another). These types are all fairly similar to the DiffTensorView.

The above are due to be consolidated into fewer types and renamed to DiffTensor

SlangPy does provide a mechanism for raw dispatch of a compute shader (see Function.dispatch), however I would recommend switching to a structure like this:

void square(int index, GradInOutTensor<float> input, GradInOutTensor<float> output)
{
    output[index] = input[index] * input[index];
}

You can then call that function, passing in a grid generator as the index (https://slangpy.shader-slang.org/en/latest/src/generators/generator_grid.html)

That all said, this stuff is in flux at the moment. Our initial release of SlangPy was very focussed on 'slangpy-like-programs' and less so on porting old code that used slang-torch. This area is something we're now actively looking at.

**Comment by l-oneil-arm** [comment]
OK, sounds great, thanks @mkeshavaNV.

**Comment by ShnitzelKiller** [comment]
Just to be sure, is the current working way to provide random access to torch tensors in slangpy to use RWTensor for everything? NOT GradInOutTensor? And without specifying `in` or `out` on the arguments? How does the automatically-created custom autograd Function know which of the arguments in @l-oneil-arm 's example are inputs and which are outputs?

**Comment by mkeshavaNV** [comment]
> Just to be sure, is the current working way to provide random access to torch tensors in slangpy to use RWTensor for everything? NOT GradInOutTensor?
> 
> Does this mean there is no working way currently to provide both random access to input/output tensors and gradient backpropagation through a slang function, e.g. if your function is not writable in broadcastable form? If this is true, only trivially-vectorizable operations (where input and output elements depend one-to-one) can be written as Torch operations, and I would think removing the recommendation at the top of the slang-torch documentation might be warranted:
> 
> > Note: This documentation is about slang-torch, an old way to use Slang with Python and PyTorch. Developers who are building new projects should use [SlangPy](https://slangpy.shader-slang.org/) instead.

@ShnitzelKiller - You should be able to use `ITensor` and slangpy internally synthesizes the right struct for you. There is a plan to revamp some of the Grad*Tensor interfaces soon, so this stuff will change going ahead.
Cc: @ccummingsNV for additonal comments.

**Comment by ShnitzelKiller** [comment]
I cannot for the life of me create a working torch op with an ITensor argument as well as a custom backwards derivative (#639), which I need for my purposes of using a stored index from the forward pass to make the backwards computation tractable.

**Comment by mkeshavaNV** [comment]
@l-oneil-arm - We've recently made some improvements to slangpy.
1. There was a huge Tensor refactor update that provides the equivalence requested in this issue.
2. slangpy also now supports `DiffTensorView `and `TensorView `data structures from slangtorch, although this is just for porting compatibility reasons. This is noted in the docs [here](https://github.com/shader-slang/slangpy/blob/034fd99ca4c340d96d88b2352a3f88c4939775b2/docs/src/autodiff/pytorch.rst)

I will go ahead and close this. Kindly re-open this if you see new issues. Thank you!
